File: src/mainview/core/toolStatus.ts
import { compareVersions, getMajor, isPreReleaseVersion } from "../../shared/version";
import { resolveBaseVersion } from "./helpers";
import type { PluginRow } from "./types";

/** Each installed major is tracked independently, including non-global versions. */
export function getInstalledSeries(plugin: PluginRow) {
	const installed = new Map<number, string>();
	for (const version of plugin.installedVersions) {
		const major = getMajor(version);
		if (major === null) continue;
		const previous = installed.get(major);
		if (!previous || compareVersions(version, previous) > 0) installed.set(major, version);
	}
	return [...installed].sort(([a], [b]) => b - a).map(([major, current]) => {
		const candidate = plugin.latestByMajor[major] ?? (getMajor(resolveBaseVersion(plugin) ?? "") === major ? plugin.sameMajorLatest : null);
		const latest = plugin.status === "done" && candidate && !isPreReleaseVersion(candidate) ? candidate : null;
		return { major, current, latest, update: latest && compareVersions(latest, current) > 0 ? latest : null };
	});
}

export function hasToolUpdates(plugin: PluginRow): boolean {
	return !!getToolUpdate(plugin) || getInstalledSeries(plugin).some(series => series.update);
}

/** Verified stable updates; prereleases remain separate in the version inspector. */
export function getToolUpdate(plugin: PluginRow): { primary: string | null; major: string | null } | null {
	const base = resolveBaseVersion(plugin);
	if (plugin.status !== "done" || !base || getMajor(base) === null) return null;
	const needsInstall = (version: string) => !plugin.installedVersions.some(installed =>
		getMajor(installed) === getMajor(version) && compareVersions(installed, version) >= 0);
	const primary = plugin.sameMajorLatest && needsInstall(plugin.sameMajorLatest) && !isPreReleaseVersion(plugin.sameMajorLatest) && compareVersions(plugin.sameMajorLatest, base) > 0
		? plugin.sameMajorLatest : null;
	const major = plugin.releaseLatest && needsInstall(plugin.releaseLatest) && !isPreReleaseVersion(plugin.releaseLatest) && getMajor(plugin.releaseLatest)! > getMajor(base)! && compareVersions(plugin.releaseLatest, base) > 0
		? plugin.releaseLatest : null;
	return primary || major ? { primary, major } : null;
}

export function getToolStatus(plugin: PluginRow): string {
	if (plugin.status === "error") return "확인 실패";
	if (plugin.status === "checking") return "확인 중";
	if (plugin.status === "updating") return "변경 중";
	if (plugin.status === "deleting") return "삭제 중";
	if (plugin.status !== "done") return "미확인";
	const base = resolveBaseVersion(plugin);
	if (getMajor(base ?? "") === null) return "채널 버전";
	if (hasToolUpdates(plugin)) return "업데이트";
	if (!plugin.sameMajorLatest && !plugin.releaseLatest) return "비교 정보 없음";
	if (isPreReleaseVersion(base!)) return "프리릴리스 사용";
	return "최신 안정 버전";
}

File: Packages/MiseCore/Sources/MiseCore/Rules/ToolStatus.swift
/// One installed major line of a tool, tracked independently of the global selection.
public struct InstalledSeries: Equatable, Sendable {
    public let major: UInt64
    public let current: String
    public let latest: String?
    public let update: String?
    public init(major: UInt64, current: String, latest: String?, update: String?) {
        self.major = major; self.current = current; self.latest = latest; self.update = update
    }
}

public struct ToolUpdate: Equatable, Sendable {
    public let primary: String?
    public let major: String?
    public init(primary: String?, major: String?) { self.primary = primary; self.major = major }
}

/// Port of src/mainview/core/toolStatus.ts.
public enum ToolStatus {
    public static func installedSeries(_ plugin: PluginRow) -> [InstalledSeries] {
        var newest: [UInt64: String] = [:]
        for version in plugin.installedVersions {
            guard let major = Version.major(of: version) else { continue }
            if let previous = newest[major], Version.compare(version, previous) != .orderedDescending { continue }
            newest[major] = version
        }
        let baseMajor = plugin.baseVersion.flatMap(Version.major(of:))
        return newest.keys.sorted(by: >).map { major in
            let current = newest[major]!
            let candidate = plugin.latestByMajor[String(major)] ?? (baseMajor == major ? plugin.sameMajorLatest : nil)
            let latest = plugin.status == .done ? candidate.flatMap { Version.isPreRelease($0) ? nil : $0 } : nil
            let update = latest.flatMap { Version.compare($0, current) == .orderedDescending ? $0 : nil }
            return InstalledSeries(major: major, current: current, latest: latest, update: update)
        }
    }

    public static func hasUpdates(_ plugin: PluginRow) -> Bool {
        update(plugin) != nil || installedSeries(plugin).contains { $0.update != nil }
    }

    /// Verified stable updates only; prereleases stay in the version inspector.
    public static func update(_ plugin: PluginRow) -> ToolUpdate? {
        guard plugin.status == .done, let base = plugin.baseVersion, let baseMajor = Version.major(of: base) else { return nil }
        func needsInstall(_ version: String) -> Bool {
            !plugin.installedVersions.contains {
                Version.major(of: $0) == Version.major(of: version) && Version.compare($0, version) != .orderedAscending
            }
        }
        func newer(_ candidate: String?) -> String? {
            candidate.flatMap {
                needsInstall($0) && !Version.isPreRelease($0) && Version.compare($0, base) == .orderedDescending ? $0 : nil
            }
        }
        let primary = newer(plugin.sameMajorLatest)
        let major = newer(plugin.releaseLatest).flatMap { Version.major(of: $0).map { $0 > baseMajor } == true ? $0 : nil }
        return primary == nil && major == nil ? nil : ToolUpdate(primary: primary, major: major)
    }

    public static func label(_ plugin: PluginRow) -> String {
        switch plugin.status {
        case .error: return Strings.ToolStatus.checkFailed
        case .checking: return Strings.ToolStatus.checking
        case .updating: return Strings.ToolStatus.changing
        case .deleting: return Strings.ToolStatus.deleting
        case .idle, .skipped: return Strings.ToolStatus.unchecked
        case .done: break
        }
        guard let base = plugin.baseVersion, Version.major(of: base) != nil else { return Strings.ToolStatus.channel }
        if hasUpdates(plugin) { return Strings.ToolStatus.updateAvailable }
        if plugin.sameMajorLatest == nil && plugin.releaseLatest == nil { return Strings.ToolStatus.noComparison }
        if Version.isPreRelease(base) { return Strings.ToolStatus.prerelease }
        return Strings.ToolStatus.latestStable
    }
}
